### PR TITLE
Add set/update extended methods

### PR DESCRIPTION
**Pull request by corrado4eyes**
#35

### CODE REVIEW

File: firebase-firestore/src/androidMain/kotlin/dev/gitlive/firebase/firestore/firestore.kt [hunk 1]
```diff
@@ -114,6 +140,27 @@ actual class WriteBatch(val android: com.google.firebase.firestore.WriteBatch) {
...
+
+        val serializedFieldAndValues = fieldsAndValues.takeUnless { fieldsAndValues.isEmpty() }
+            ?.map { (field, value) -> field to encode(value, encodeDefaults) }?.toMap()
+
+        val result = (serializedFieldAndValues?.let {
+            serializedItem.plus(it)
+        } ?: serializedItem)
+
+        return android.update(documentRef.android, result).let { this }
+    }
```
Comment (ChristoferAlexander) [inline firebase-firestore/src/androidMain/kotlin/dev/gitlive/firebase/firestore/firestore.kt:161]: Android specific impl of new added method, basically it combines the mapping of data and fieldsAndValues into one update query. Use case example is an update on a data class A with fields A.field1 and A.field2 where field1 value needs to be updated and field2 value needs to be deleted. In that case new value for field1 is included on the "data" and field2 in the fieldsAndValues. (Due to limitation of FieldValue.delete() or FieldValue.serverTimestamp() that can only by included in fieldsAndValues and not in the data class itself)

File: firebase-firestore/src/iosMain/kotlin/dev/gitlive/firebase/firestore/firestore.kt [hunk 2]
```diff
@@ -102,7 +122,23 @@ actual class WriteBatch(val ios: FIRWriteBatch) {
...
+        vararg fieldsAndValues: Pair<String, Any?>
+    ): WriteBatch {
+        val serializedItem = encode(strategy, data, encodeDefaults) as Map<Any?, *>
+        val serializedFieldAndValues = fieldsAndValues.associate { (field, value) -> field to encode(value, encodeDefaults) }
+
+        val result = serializedFieldAndValues?.let {
+            serializedItem.plus(it)
+        } ?: serializedItem
+        return ios.updateData(result, documentRef.ios).let { this }
+    }
```
Comment (ChristoferAlexander) [inline firebase-firestore/src/iosMain/kotlin/dev/gitlive/firebase/firestore/firestore.kt:139]: Same as Android

File: firebase-firestore/src/androidMain/kotlin/dev/gitlive/firebase/firestore/firestore.kt [hunk 3]
```diff
@@ -99,6 +99,32 @@ actual class WriteBatch(val android: com.google.firebase.firestore.WriteBatch) {
...
+            }
+        } as Any? ?: return this
+
+        if (merge) {
+            android.set(documentRef.android, result, SetOptions.merge())
+        } else {
+            android.set(documentRef.android, result)
+        }
+        return this
+    }
```
Comment (ChristoferAlexander) [inline firebase-firestore/src/androidMain/kotlin/dev/gitlive/firebase/firestore/firestore.kt:126]: Same as update bellow.

File: firebase-firestore/src/iosMain/kotlin/dev/gitlive/firebase/firestore/firestore.kt [hunk 4]
```diff
@@ -90,6 +90,26 @@ actual class WriteBatch(val ios: FIRWriteBatch) {
...
+        val serializedFieldAndValues = fieldsAndValues.takeUnless { fieldsAndValues.isEmpty() }
+            ?.map { (field, value) -> field to encode(value, encodeDefaults) }?.toMap()
+
+        val result = (serializedFieldAndValues?.let {
+            serializedItem?.plus(it)
+        } ?: serializedItem) as Map<Any?, *>
+
+        ios.setData(result, documentRef.ios, merge)
+        return this
+    }
```
Comment (ChristoferAlexander) [inline firebase-firestore/src/iosMain/kotlin/dev/gitlive/firebase/firestore/firestore.kt:111]: Same as update bellow

File: firebase-firestore/src/jsMain/kotlin/dev/gitlive/firebase/firestore/firestore.kt [hunk 5]
```diff
@@ -86,6 +86,32 @@ actual class WriteBatch(val js: firebase.firestore.WriteBatch) {
...
+            }
+        } as Any? ?: return this
+
+        if (merge) {
+            js.set(documentRef.js, result, json("merge" to merge))
+        } else {
+            js.set(documentRef.js, result)
+        }
+        return this
+    }
```
Comment (ChristoferAlexander) [inline firebase-firestore/src/jsMain/kotlin/dev/gitlive/firebase/firestore/firestore.kt:113]: Same for JS

File: firebase-firestore/src/jsMain/kotlin/dev/gitlive/firebase/firestore/firestore.kt [hunk 6]
```diff
@@ -100,6 +126,25 @@ actual class WriteBatch(val js: firebase.firestore.WriteBatch) {
...
+
+        val serializedFieldAndValues = fieldsAndValues.takeUnless { fieldsAndValues.isEmpty() }
+            ?.map { (field, value) -> field to encode(value, encodeDefaults) }?.toMap()
+
+        val result = (serializedFieldAndValues?.let {
+            serializedItem.plus(it)
+        } ?: serializedItem)
+
+        return js.update(documentRef.js, result).let { this }
+    }
```
Comment (ChristoferAlexander) [inline firebase-firestore/src/jsMain/kotlin/dev/gitlive/firebase/firestore/firestore.kt:146]: Same for JS

**Review by ChristoferAlexander — APPROVED** [review]

File: firebase-firestore/src/iosMain/kotlin/dev/gitlive/firebase/firestore/firestore.kt [hunk 7]
```diff
@@ -90,6 +90,26 @@ actual class WriteBatch(val ios: FIRWriteBatch) {
...
+        strategy: SerializationStrategy<T>,
+        data: T,
+        encodeDefaults: Boolean,
+        merge: Boolean,
+        vararg fieldsAndValues: Pair<String, Any?>
+    ): WriteBatch {
+        val serializedItem = encode(strategy, data, encodeDefaults) as Map<String, Any>?
+        val serializedFieldAndValues = fieldsAndValues.takeUnless { fieldsAndValues.isEmpty() }
+            ?.map { (field, value) -> field to encode(value, encodeDefaults) }?.toMap()
+
```
Comment (ChristoferAlexander) [inline firebase-firestore/src/iosMain/kotlin/dev/gitlive/firebase/firestore/firestore.kt:104]:
```suggestion
```

File: firebase-firestore/src/jsMain/kotlin/dev/gitlive/firebase/firestore/firestore.kt (hunk 6 above)
Comment (corrado4eyes) [reply firebase-firestore/src/jsMain/kotlin/dev/gitlive/firebase/firestore/firestore.kt]: c5e313c 6e74c62 c9e98ce 30fc496

File: firebase-firestore/src/jsMain/kotlin/dev/gitlive/firebase/firestore/firestore.kt (hunk 5 above)
Comment (corrado4eyes) [reply firebase-firestore/src/jsMain/kotlin/dev/gitlive/firebase/firestore/firestore.kt]: c5e313c 6e74c62 c9e98ce 30fc496

File: firebase-firestore/src/androidMain/kotlin/dev/gitlive/firebase/firestore/firestore.kt (hunk 3 above)
Comment (corrado4eyes) [reply firebase-firestore/src/androidMain/kotlin/dev/gitlive/firebase/firestore/firestore.kt]: c5e313c 6e74c62 c9e98ce 30fc496

File: firebase-firestore/src/iosMain/kotlin/dev/gitlive/firebase/firestore/firestore.kt (hunk 2 above)
Comment (corrado4eyes) [reply firebase-firestore/src/iosMain/kotlin/dev/gitlive/firebase/firestore/firestore.kt]: c5e313c 6e74c62 c9e98ce 30fc496

File: firebase-firestore/src/androidMain/kotlin/dev/gitlive/firebase/firestore/firestore.kt (hunk 1 above)
Comment (corrado4eyes) [reply firebase-firestore/src/androidMain/kotlin/dev/gitlive/firebase/firestore/firestore.kt]: c5e313c 6e74c62 c9e98ce 30fc496

**Pull request merged**